### PR TITLE
Feature: Add GPU memory information in arena top node.

### PR DESCRIPTION
**Pull request by Sakuralbj**
## Premise: GPU resources on a node are either allocated by card or distributed by memory.

1.arena top node : Show the GPU resource usage of all nodes，If there is a GPUShare node in the entire cluster, the node is displayed as shareable.

2.arena top node -s : Show the GPUMemory resource usage of all GPUShare nodes.

3.arena top node -d: Show details of GPUCount task on a node. Including the name of the task, namespace, and GPUCount used.

4.arena top node -s -d:Show details of GPUMemory task on a GPUShare node. Including the name of the task, namespace, and GPUMemory used.  

## Test Case:
Node cn-shanghai.i-uf61h64dz1tmlob9hmtb is GPUShare, one card ,aliyun.com/gpu-mem=15.
Node cn-shanghai.i-uf61h64dz1tmlob9hmtc is GPUCount, one card, nvidia.com/gpu=1.
Now I create two serving tasks, mymnist1 request nvidia.com/gpu=1,mymnist2 requests aliyun.com/gpu-mem=3.
### Device-plugin
```
[root@iZuf69zddmom136duk79quZ arena]# kubectl get po -o wide -n kube-system|grep device-plugin
gpushare-device-plugin-ds-4src6                              1/1     Running       0          3d5h   192.168.0.71   cn-shanghai.i-uf61h64dz1tmlob9hmtb   <none>
nvidia-device-plugin-cn-shanghai.i-uf61h64dz1tmlob9hmtc      1/1     Running       0          86d    192.168.0.70   cn-shanghai.i-uf61h64dz1tmlob9hmtc   <none>
```
### arena serve list
```
[root@iZuf69zddmom136duk79quZ arena]# arena serve list
NAME      TYPE        VERSION  DESIRED  AVAILABLE  ENDPOINT_ADDRESS  PORTS
mymnist1  TENSORFLOW           1        1          172.19.15.32      serving:8500,http-serving:8501
mymnist2  TENSORFLOW           1        1          172.19.7.104      serving:8500,http-serving:8501
```
### arena top node . 
```
[root@iZuf69zddmom136duk79quZ arena]# arena top node
NAME                                IPADDRESS     ROLE    STATUS  GPU(Total)  GPU(Allocated)  GPUShare
cn-shanghai.i-uf61h64dz1tmlob9hmtb  192.168.0.71  <none>  ready   1           1               Shareable
cn-shanghai.i-uf61h64dz1tmlob9hmtc  192.168.0.70  <none>  ready   1           1               N/A
cn-shanghai.i-uf6347ba9krw8hj5yvsy  192.168.0.67  master  ready   0           0               N/A
cn-shanghai.i-uf662a07bhojl329pity  192.168.0.68  master  ready   0           0               N/A
cn-shanghai.i-uf69zddmom136duk79qu  192.168.0.69  master  ready   0           0               N/A
------------------------------------------------------------------------------------------------
Allocated/Total GPUs In Cluster:
2/2 (100%)  
```
### arena top node -d . 
```
NAME:                                                       cn-shanghai.i-uf61h64dz1tmlob9hmtb
IPADDRESS:                                                  192.168.0.71
ROLE:                                                       <none>
Total GPUs In Node cn-shanghai.i-uf61h64dz1tmlob9hmtb:      1         
Allocated GPUs In Node cn-shanghai.i-uf61h64dz1tmlob9hmtb:  1 (100%)  
If Node is Shareable :                                      Yes 
-----------------------------------------------------------------------------------------

NAME:       cn-shanghai.i-uf61h64dz1tmlob9hmtc
IPADDRESS:  192.168.0.70
ROLE:       <none>

NAMESPACE  NAME                                          GPU REQUESTS  GPU LIMITS
default    mymnist1-tensorflow-serving-76d5c7c8fc-nbmtk  1             1

Total GPUs In Node cn-shanghai.i-uf61h64dz1tmlob9hmtc:      1         
Allocated GPUs In Node cn-shanghai.i-uf61h64dz1tmlob9hmtc:  1 (100%)  
-----------------------------------------------------------------------------------------

NAME:       cn-shanghai.i-uf6347ba9krw8hj5yvsy
IPADDRESS:  192.168.0.67
ROLE:       master

Total GPUs In Node cn-shanghai.i-uf6347ba9krw8hj5yvsy:      0       
Allocated GPUs In Node cn-shanghai.i-uf6347ba9krw8hj5yvsy:  0 (0%)  
-----------------------------------------------------------------------------------------

NAME:       cn-shanghai.i-uf662a07bhojl329pity
IPADDRESS:  192.168.0.68
ROLE:       master

Total GPUs In Node cn-shanghai.i-uf662a07bhojl329pity:      0       
Allocated GPUs In Node cn-shanghai.i-uf662a07bhojl329pity:  0 (0%)  
-----------------------------------------------------------------------------------------

NAME:       cn-shanghai.i-uf69zddmom136duk79qu
IPADDRESS:  192.168.0.69
ROLE:       master

Total GPUs In Node cn-shanghai.i-uf69zddmom136duk79qu:      0       
Allocated GPUs In Node cn-shanghai.i-uf69zddmom136duk79qu:  0 (0%)  
-----------------------------------------------------------------------------------------


Allocated/Total GPUs In Cluster:  2/2 (100%)  
```
### arena top node -s . 
```
[root@iZuf69zddmom136duk79quZ arena]# arena top node -s
NAME                                IPADDRESS     GPU0(Allocated/Total)  
cn-shanghai.i-uf61h64dz1tmlob9hmtb  192.168.0.71  3/15                   
--------------------------------------------------------------------------
Allocated/Total GPU Memory In Cluster:
3/15 (GiB) (20%)  
```
### arena top node -s -d
```
[root@iZuf69zddmom136duk79quZ arena]# arena top node -s -d

NAME:       cn-shanghai.i-uf61h64dz1tmlob9hmtb
IPADDRESS:  192.168.0.71

NAME                                          NAMESPACE  GPU0(Allocated)  
mymnist2-tensorflow-serving-7446c98547-87bqh  default    3                
Allocated :                                   3 (20%)    
Total :                                       15         
------------------------------------------------------------------


Allocated/Total GPU Memory In Cluster:
3/15 (GiB) (20%)  
```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/arena/226)

### CODE REVIEW

**Comment by Sakuralbj** [comment]
I have updated arena top node -d. This is new test result.
```
#arena serve list
NAME      TYPE        VERSION  DESIRED  AVAILABLE  ENDPOINT_ADDRESS  PORTS
mymnist1  TENSORFLOW           1        1          172.19.9.182      serving:8500,http-serving:8501
mymnist2  TENSORFLOW           1        1          172.19.9.194      serving:8500,http-serving:8501
mymnist3  TENSORFLOW           1        1          172.19.10.191     serving:8500,http-serving:8501
```
```
# arena top node 
NAME                                IPADDRESS     ROLE    STATUS  GPU(Total)  GPU(Allocated)  GPU(Shareable)
cn-shanghai.i-uf61h64dz1tmlob9hmtb  192.168.0.71  <none>  ready   1           1               Yes
cn-shanghai.i-uf61h64dz1tmlob9hmtc  192.168.0.70  <none>  ready   1           1               No
cn-shanghai.i-uf6347ba9krw8hj5yvsy  192.168.0.67  master  ready   0           0               No
cn-shanghai.i-uf662a07bhojl329pity  192.168.0.68  master  ready   0           0               No
cn-shanghai.i-uf69zddmom136duk79qu  192.168.0.69  master  ready   0           0               No
------------------------------------------------------------------------------------------------
Allocated/Total GPUs In Cluster:
2/2 (100%)  
```
```
#arena top node -d
NAME:       cn-shanghai.i-uf61h64dz1tmlob9hmtb
IPADDRESS:  192.168.0.71
ROLE:       <none>

NAMESPACE  NAME                                          GPU REQUESTS   
default    mymnist2-tensorflow-serving-7446c98547-snglf  0.20          
default    mymnist3-tensorflow-serving-787689ffd-69r66   0.20          

Total GPUs In Node cn-shanghai.i-uf61h64dz1tmlob9hmtb:      1         
Allocated GPUs In Node cn-shanghai.i-uf61h64dz1tmlob9hmtb:  1 (100%)  
If Node is Shareable :                                      Yes 
-----------------------------------------------------------------------------------------

NAME:       cn-shanghai.i-uf61h64dz1tmlob9hmtc
IPADDRESS:  192.168.0.70
ROLE:       <none>

NAMESPACE  NAME                                          GPU REQUESTS   
default    mymnist1-tensorflow-serving-76d5c7c8fc-pld9h  1             

Total GPUs In Node cn-shanghai.i-uf61h64dz1tmlob9hmtc:      1         
Allocated GPUs In Node cn-shanghai.i-uf61h64dz1tmlob9hmtc:  1 (100%)  
-----------------------------------------------------------------------------------------

NAME:       cn-shanghai.i-uf6347ba9krw8hj5yvsy
IPADDRESS:  192.168.0.67
ROLE:       master

Total GPUs In Node cn-shanghai.i-uf6347ba9krw8hj5yvsy:      0       
Allocated GPUs In Node cn-shanghai.i-uf6347ba9krw8hj5yvsy:  0 (0%)  
-----------------------------------------------------------------------------------------

NAME:       cn-shanghai.i-uf662a07bhojl329pity
IPADDRESS:  192.168.0.68
ROLE:       master

Total GPUs In Node cn-shanghai.i-uf662a07bhojl329pity:      0       
Allocated GPUs In Node cn-shanghai.i-uf662a07bhojl329pity:  0 (0%)  
-----------------------------------------------------------------------------------------

NAME:       cn-shanghai.i-uf69zddmom136duk79qu
IPADDRESS:  192.168.0.69
ROLE:       master

Total GPUs In Node cn-shanghai.i-uf69zddmom136duk79qu:      0       
Allocated GPUs In Node cn-shanghai.i-uf69zddmom136duk79qu:  0 (0%)  
-----------------------------------------------------------------------------------------


Allocated/Total GPUs In Cluster:  2/2 (100%)  
```
@cheyang

File: cmd/arena/commands/top_node.go [hunk 1]
```diff
@@ -62,11 +64,47 @@ func NewTopNodeCommand() *cobra.Command {
 				fmt.Println(err)
 				os.Exit(1)
 			}
+			tmp := false
```
Comment (cheyang) [inline cmd/arena/commands/top_node.go]: we need the meaningful variable definition here.

Comment (Sakuralbj) [reply cmd/arena/commands/top_node.go]: Thanks for your guidance, I have modified these places, please check it again.

File: cmd/arena/commands/top_node.go [hunk 2]
```diff
@@ -62,11 +64,32 @@ func NewTopNodeCommand() *cobra.Command {
...
 			}
-			displayTopNode(nodeInfos)
+			for _, nodeInfo := range nodeInfos {
+				if gpushare.IsGPUShareNode(nodeInfo.node) && isGPUCountNode(nodeInfo.node) {
+					fmt.Printf("nvidia.com/gpu and aliyun.com/gpu-mem are both used in node %s .Please use one kind resource of them\n", nodeInfo.node.Name)
+					os.Exit(1)
+				}
+			}
+			if showGPUShare {
+				gpushareNodes, err := GetAllGPUShareNode()
```
Comment (cheyang) [inline cmd/arena/commands/top_node.go]: This operation will fetch all the nodes again. Can you enhance it?

File: cmd/arena/commands/const.go [hunk 3]
```diff
@@ -19,6 +19,7 @@ const (
 	// GPUResourceName is the extended name of the GPU resource since v1.8
 	// this uses the device plugin mechanism
 	NVIDIAGPUResourceName = "nvidia.com/gpu"
+	resourceName          = "aliyun.com/gpu-mem"
```
Comment (cheyang) [inline cmd/arena/commands/const.go]: Could you follow the Convention？

File: cmd/arena/commands/top_node.go [hunk 4]
```diff
@@ -57,16 +59,32 @@ func NewTopNodeCommand() *cobra.Command {
 				os.Exit(1)
 			}
 			nd := newNodeDescriber(client, allPods)
-			nodeInfos, err := nd.getAllNodeInfos()
+			nodeInfos, gpushareNodes, err := nd.getAllNodeInfos()
```
Comment (cheyang) [inline cmd/arena/commands/top_node.go]: Please don't add `gpushareNodes` which may confuse others.  I think `isGpushare` is an attribute of `nodeInfo`.

File: cmd/arena/commands/top_node.go [hunk 5]
```diff
@@ -201,7 +207,7 @@ func displayTopNodeSummary(nodeInfos []NodeInfo) {
 		var allocatableGPU int64
 		var allocatedGPU int64
 		// GPUShare nodes and normal nodes  calculate the allocatedGPU and total GPU in different way
-		if isGPUShareNode(nodeInfo) {
+		if nodeInfo.isGpushare {
```
Comment (cheyang) [inline cmd/arena/commands/top_node.go]: it should change to `isGPUShareNode ` to keep consistent.

**Pull request merged**